platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/ServiceScopeValidator.java

```diff
@@ -52,24 +52,52 @@ public void whenRegistered(Class<? extends Annotation> annotation, Registration
     }
 
     private void validateScope(Class<?> serviceType) {
-        Class<? extends Scope> serviceScope = scopeOf(serviceType);
+        Class<? extends Scope>[] serviceScopes = scopeOf(serviceType);
 
-        if (serviceScope == null || ServiceScopeValidatorWorkarounds.shouldSuppressValidation(serviceType)) {
+        if (serviceScopes == null) {
             return;
         }
 
-        if (!scope.equals(serviceScope)) {
-            throw new IllegalArgumentException(invalidScopeMessage(serviceType, serviceScope));
+        if (serviceScopes.length == 0) {
+            throw new IllegalArgumentException(String.format("Service '%s' is declared with empty scope list", serviceType.getName()));
+        }
+
+        if (!containsScope(scope, serviceScopes)) {
+            throw new IllegalArgumentException(invalidScopeMessage(serviceType, serviceScopes));
         }
     }
 
-    private String invalidScopeMessage(Class<?> serviceType, Class<? extends Scope> actualScope) {
-        return String.format("Service '%s' was declared in scope '%s' but registered in scope '%s'",
-            serviceType.getName(), actualScope.getSimpleName(), scope.getSimpleName());
+    private static boolean containsScope(Class<? extends Scope> expectedScope, Class<? extends Scope>[] actualScopes) {
+        for (Class<? extends Scope> declScope : actualScopes) {
+            if (expectedScope.equals(declScope)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private String invalidScopeMessage(Class<?> serviceType, Class<? extends Scope>[] actualScopes) {
+        return String.format("Service '%s' was declared in scopes %s but registered in scope '%s'",
+            serviceType.getName(), scopeDisplayList(actualScopes), scope.getSimpleName());
+    }
+
+    private static String scopeDisplayList(Class<? extends Scope>[] scopes) {
+        StringBuilder sb = new StringBuilder();
```

Comment:
💅 Extract `org.gradle.internal.util.ArrayExtensions.joinToString<T>(T[], (T) -> String)`

Reply:
I found some ready utilities in the `CollectionsUtils`. I hope the code is now clearer.